doc/rewriting-services.md

```diff
@@ -0,0 +1,204 @@
+update the tests _and_ the examples. Take care to do that.
+
+## Implement `render()` and `handle()`
+
+1. Once the route is working, fill out `render()` and `handle()`. If there’s a
+single service, you can implement fetch as a method or a function at the top of
+the file. If there are more than one service which share fetching code, you can
+put the fetch function in `example-common.js` like this one for github:
+
+<details>
+
+```js
+const Joi = require('joi')
+const { errorMessagesFor } = require('./github-helpers')
+
+const issueSchema = Joi.object({
+  head: Joi.object({
+    sha: Joi.string().required(),
+  }).required(),
+}).required()
+
+async function fetchIssue(serviceInstance, { user, repo, number }) {
+  return serviceInstance._requestJson({
+    schema: issueSchema,
+    url: `/repos/${user}/${repo}/pulls/${number}`,
+    errorMessages: errorMessagesFor('pull request or repo not found'),
+  })
+}
+
+module.exports = {
+  fetchIssue,
+}
+```
+
+</details>
+
+or create an abstract superclass like **PypiBase**:
+
+<details>
+
+```js
+const Joi = require('joi')
+const BaseJsonService = require('../base-json')
+
+const schema = Joi.object({
+  info: Joi.object({
+    version: Joi.string().required(),
+    // https://github.com/badges/shields/issues/2022
+    license: Joi.string().allow(''),
+    classifiers: Joi.array()
+      .items(Joi.string())
+      .required(),
+  }).required(),
+  releases: Joi.object()
+    .pattern(
+      Joi.string(),
+      Joi.array()
+        .items(
+          Joi.object({
+            packagetype: Joi.string().required(),
+          })
+        )
+        .required()
+
+module.exports = class PypiBase extends BaseJsonService {
+  static buildRoute(base) {
+    return {
+      base,
+      pattern: ':egg*',
+    }
+  }
+
+  async fetch({ egg }) {
+    return this._requestJson({
+      schema,
+      url: `https://pypi.org/pypi/${egg}/json`,
+      errorMessages: { 404: 'package or version not found' },
+    })
+  }
+}
+```
+
+</details>
+
+2. To keep with the design pattern of `render()`, formatting concerns, including
+concatenation and color computation, should be dealt with inside `render()`.
+This helps avoid static examples falling out of sync with the implementation.
+
+## Convert the examples
+
+1. Convert all the examples to `pattern`, `namedParams`, and `staticExample`. In some cases you can use the `pattern` inherited from `route`, though in other cases you may need to specify a pattern in the example. For example, when showing download badges for several periods, you may want to render the example with an explicit `dt` instead of `:which`. You will also need to specify a pattern for badges that use a `format` regex in the route.
+
+2. Open the frontend and check that the static preview badges look good.
+Remember, none of them are live.
+
+3. Open up the prepared example URLs in their own tabs, and make sure they work correctly.
+
+## Validation
+
+Add the schemas toward the end.
```

Comment:
If we're going to suggest adding the schema at the end, we should clarify that you can use `Joi.any()` to start with. We usually need to pass _some_ schema at an earlier stage than this.

Also from the perspective of refactoring, we should explain something about converting manual validation rules to schema. A typical legacy service will have some `if (myvar !== undefined)...` or `const somevar = myvar || 'default'` etc that we want to convert to schema validation.